test: add comprehensive unit tests for HabitAssignController

MykytaSirobaba opened the pull request:
- Covered all key endpoints: POST, PUT/PATCH, GET, DELETE
- Verified correct handling of JSON inputs, path/query params, and localization (Locale)
- Mocked HabitAssignService and @CurrentUser (UserVO)
- Used @ExtendWith(MockitoExtension.class) for unit test structure
- Ensured DTO-based response validation and service interaction via Mockito.verify(...)
- Achieved 100% test coverage on controller methods

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Tests**
  * Added comprehensive tests covering all habit assignment management features, ensuring reliable handling of habit assignments, updates, enrollments, cancellations, and progress notifications.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MykytaSirobaba committed:
test: add comprehensive unit tests for HabitAssignController

- Covered all key endpoints: POST, PUT/PATCH, GET, DELETE
- Verified correct handling of JSON inputs, path/query params, and localization (Locale)
- Mocked HabitAssignService and @CurrentUser (UserVO)
- Used @ExtendWith(MockitoExtension.class) for unit test structure
- Ensured DTO-based response validation and service interaction via Mockito.verify(...)
- Achieved 100% test coverage on controller methods

MykytaSirobaba committed:
corrected some inaccuracies

natri0 commented on core/src/test/java/greencity/controller/HabitAssignControllerTest.java:
```diff
@@ -0,0 +1,483 @@
+    }
+
+    private UserVO prepareMockUser() {
+        UserVO userVO = UserVO.builder()
+                .id(1L)
+                .name("Tom")
+                .email(testEmail)
+                .build();
```
you can use the `greencity.ModelUtils` class to not have to create own `UserVO`, but this is just a minor nitpick

MykytaSirobaba replied:
If we use it, we have to write when(userService.findByEmail(testEmail)).thenReturn(userVO); , in all methods when we have a UserVO, but with this method I can avoid it; However this is a good point.

natri0 commented on core/src/test/java/greencity/controller/HabitAssignControllerTest.java:
```diff
@@ -0,0 +1,483 @@
+                  },
+                  "friendsIdsList": [2, 3, 4]
+                }
+                """;
```
you're already using a `ObjectMapper` in some other places, why not here as well?

MykytaSirobaba replied:
I absolutely agree with you. I forget about it.

MykytaSirobaba committed:
minor inaccuracies were fixed

AndriiSynytsia commented on core/src/test/java/greencity/controller/HabitAssignControllerTest.java:
```diff
@@ -0,0 +1,478 @@
+
+    @Test
+    void assignDefault() throws Exception {
+        Long habitId = 1L;
+        UserVO userVO = ModelUtils.getUserVO();
```
Hey. You did great job. I checked every single line of code and I noticed that those two lines repeats every single time. Would be nice to have it as constants in this test class. Not sure but there could be couple more improvements repetitive code

MykytaSirobaba replied:
We should run all methods independently, because if one of them changes its behavior or modifies some fields, it could cause an error. Currently, final initializes all methods at once and uses the same user for all of them.

AndriiSynytsia commented on core/src/test/java/greencity/controller/HabitAssignControllerTest.java:
```diff
@@ -0,0 +1,478 @@
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(objectMapper.writeValueAsString(habitAssignManagementDto))
+                        .accept(MediaType.APPLICATION_JSON))
+                .andExpect(status().isCreated());
```
When you sending post request create the return type is HabitAssignDto that would be great to assert with example. That you could apply to the rest of tests

AndriiSynytsia reviewed (changes requested):
The code looks nice and clean. There were couple comments about repetitive code and checking assertions. Also according to pretty common naming of test cases you could use DisplayName, it is not necessary but it is a good practice to explain people who read your code what it actually do

MykytaSirobaba committed:
Add response content assertion to assignCustom test

- Added check for JSON response content in assignCustom() test
- Ensure the returned DTO list matches expected output
